CIPs/0017.md

```diff
@@ -0,0 +1,69 @@
+
+## Goals
+
+- Support adding/storing proofs on full nodes
+- Gossip these proofs to propogate them through the network
+- Extend the `les` protocol such that the server will provide these proofs in place of headers (when appropriate), and the client will consume/verify proofs accordingly.
+
+## Proposed Solution
+
+We propose to add two new RPCs: 
+`AddProof(proof []byte, proofId uint)`
+and
+`Proofs() ([][]byte)`. When an eventual "proving service" generates a new proof it will publish this proof to a full node that it is running, which will then propogate the proof throughout the network.
+
+This `AddProof` function will store the proof in a new database `plumoproofdata`, keyed by a unique proof ID.
+
+To propogate these proofs throughout the network, we also propose to extend the `eth` protocol with two new message types: 
+```
+PlumoProofMsg - Sends plumo proof(s) either in response to a `GetPlumoProofMsg` or after a successful RPC call to `AddProof`
+GetPlumoProofMsg - Request plumo proof(s) from a peer
+```
+`GetPlumoProofMsg` will be implemented by having the requester send a set of proof IDs that they hold, and ask for any they are missing. This may be optimized using a bloom filter as well.
+
+The `les` protocol must then also be extended via at least two messages:
+```
+GetEpochsMsg - Sends the current highest epoch known to the sender. Server responds with the minimal set of proofs & headers to give the light client the fastest sync time.
+EpochsMsg - Receives a mix of header(s) and proof(s) from the server.
+```
+
+The `les` client must then be modified to send a `GetEpochsMsg` message during `lightest` sync. This will likely replace the function `getEpochOrNormalHeaders` within `eth/downloader`.
+
+Upon receiving the `EpochsMsg` response message, the light client must then verify headers and proofs provided. Proofs will be verified via a call into `celo-bls-go`'s `verify` fn.
+
+The verify function requires information on an epoch's validator set, so this will likely require some access changes (private -> public) and exposing more data structures to allow the validator set to be queried and modified upon proof verification.
+
+## Alternative Solutions
+
+`GetPlumoProofMsg` could also request *all* plumo proofs from the receiver, leaving the responsibility of filtering seen proofs with the caller. This is a simpler implementation, but is less scalable as we increase the amount of stored proofs. 
+
+On the type of ID to use for proofs: an incrementing counter seems most "user-friendly" because of the ability to sense gaps, but could introduce some synchronization issues in the future.
+
+Specific aspects of the `les` changes may have tradeoffs that I've yet to encounter, so any input and feedback on things to consider there would be helpful.
+
+## Risks
+
+Altering the `eth` protocol introduces some risk of introducing bugs.
+
+Introducing bugs to mobile client syncing.
```

Comment:
```suggestion
Introducing bugs to mobile client syncing. Possible effects of a faulty implementation include not being able to sync to the network, and introducing a security vulnerability which would allow the node to sync to an attack controlled chain.
```